cluster-essentials/pkg/client/install.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package client has APIs for bootstrapping tanzu cluster-essentials on a kubernetes cluster.
+package client
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"k8s.io/client-go/rest"
+)
+
+var downloadBundles = downloadBundle
+var appplyResourcesFromManifests = appplyResourcesFromManifest
+
+const defaultBundleDir = "/tmp"
+const defaultTimeout = 15 * time.Minute
+
+func validatePreInstall(restConfig *rest.Config) error {
+	// TODO implementation
+	return nil
+}
+
+func validatePostInstall(restConfig *rest.Config, timeout time.Duration) error {
+	// TODO implementation
+	return nil
+}
+
+// Install installs Cluster Essentials packages.
+// This function attempts to install packages from the given imgpkg bundle path
+// and waits until install is complete or the given timeout is reached, whichever
+// occurs first.
+func Install(ctx context.Context, config *rest.Config, imgpkgBundlePath string, timeout time.Duration, bundleDestDir string) error {
+	if bundleDestDir == "" {
+		bundleDestDir = defaultBundleDir
+	}
+	if timeout == time.Duration(0) {
+		timeout = defaultTimeout
+	}
+	bundleDir, err := os.MkdirTemp(bundleDestDir, "cluster-essential")
+	defer os.RemoveAll(bundleDir)
+	if err != nil {
+		return err
+	}
+	err = downloadBundles(imgpkgBundlePath, bundleDir)
+	if err != nil {
+		return fmt.Errorf("unable to download cluster essential manifest: %w", err)
+	}
+
+	err = validatePreInstall(config)
+	if err != nil {
+		return fmt.Errorf("pre install validation is failed with error: %w", err)
+	}
+	packages, err := os.ReadDir(bundleDir)
+	if err != nil || len(packages) == 0 {
+		return fmt.Errorf("error while reading the bundleDir %s", bundleDir)
+	}
+	var generatedManifestBytes []byte
+	for _, item := range packages {
+		packageDir := filepath.Join(bundleDir, item.Name())
```

Comment:
i'm confused, you are saying this `Install` function, expects a imgpkg bundle path, but then here you say each package contains a .imgpkg directory, how can a bundle contain multiple .imgpkg directories? how does the structure of imgpkg bundle look like? may be tree structure in the documentation helps

Reply:
tree bundle/
bundle/
├── kapp-controller
│   └── config
│       ├── overlays
│       │   ├── change-namespace.yaml
│       │   ├── update-clusterrole-with-psp.yaml
│       │   ├── update-configmap.yaml
│       │   ├── update-crds.yaml
│       │   ├── update-deployment.yaml
│       │   └── update-strategy-overlay.yaml
│       ├── upstream
│       │   └── kapp-controller.yaml
│       ├── values.star
│       └── values.yaml
└── secretgen-controller
    └── config
        └── secretgen-controller.yaml

Reply:
@yharish991  each package is bundle itself